### PR TITLE
Add option to co-factor implicit data in explicit feedback

### PR DESCRIPTION
**Pull request by david-cortes**
This PR adds a very simplified version of the collective/co-factoring technique used in e.g. [this](http://cseweb.ucsd.edu/classes/fa17/cse291-b/reading/p59-liang.pdf) and similar papers for the explicit feedback model.

What it does is it basically factorizes both the explicit feedback matrix and an implicit-feedback version of it with only binary entries without weights, sharing the same components between factorizations.

These are not counted towards the loss, because it wouldn't be computationally efficient to add them as it would imply iterating over every entry of the ratings matrix, whether it's missing or not.

The procedure is however not very efficient, especially when using floats since it has the solver done in R and needs to cast at each iteration. I wasn't sure how to do it in armadillo so perhaps there's a better option for floats, or could be done directly with blas and lapack.

This PR includes the commits from the previous one, so let's merge the earlier one first.

### CODE REVIEW

File: src/SolverAlsWrmf.cpp [hunk 1]
```diff
@@ -147,11 +193,51 @@ T als_explicit(const dMappedCSC& Conf,
       // X_nnz = [x_bias, ..., 1]
       // Y_new should be [..., y_bias]
       if (solver == CONJUGATE_GRADIENT) {
-        Y_new = cg_solver_explicit<T>(X_nnz, confidence, init, lambda_use, cg_steps);
+        if (!with_implicit_features) {
+          if (with_biases) {
+            auto init = drop_row<T>(Y.col(i), !is_x_bias_last_row);
```
Comment (dselivanov) [inline src/SolverAlsWrmf.cpp:198]: I'm a bit afraid of using `auto` with arma objects. I had some segfaults and don't use `auto` with arma structures anymore. Better to use `const arma::Mat<T>` here.
In fact [arma docs](http://arma.sourceforge.net/faq.html) discourage using `auto`:
> Use of C++11 auto is not recommended with Armadillo objects and expressions.

File: src/SolverAlsWrmf.cpp [hunk 2]
```diff
@@ -147,11 +193,51 @@ T als_explicit(const dMappedCSC& Conf,
...
+            Y_new = cg_solver_explicit<T>(X_nnz, confidence, init, lambda_use, cg_steps);
+          } else {
+            Y_new = cg_solver_explicit<T>(X_nnz, confidence, Y.col(i), lambda_use, cg_steps);
+          }
+        } else {
+          if (with_biases) {
+            auto init = drop_row<T>(Y.col(i), !is_x_bias_last_row);
```
Comment (dselivanov) [inline src/SolverAlsWrmf.cpp:205]: Same `const arma::Mat<T>`

**Review by dselivanov — COMMENTED** [review]
It will take some time to review/understand underlying co-factor model. Would be great if you could add some description on when does it worth to use co-factor model. And what kind of accuracy gains one can expect.

**Comment by david-cortes** [comment]
Tried switching everything to armadillo, but now I find that it actually produces worse results compared to not using implicit features. Works quite fine in `cmfrec` though. Will close it as I have no idea how to "fix" it.

**Pull request closed without merging**